### ISSUE TITLE
🐛 <TITLE>

### ISSUE THREAD

**Issue by mike-baton** [biomejs/biome#5901]
### Environment information

```block
CLI:
  Version:                      2.0.0-beta.2
  Color support:                true

Platform:
  CPU Architecture:             aarch64
  OS:                           macos

Environment:
  BIOME_LOG_PATH:               unset
  BIOME_LOG_PREFIX_NAME:        unset
  BIOME_CONFIG_PATH:            unset
  NO_COLOR:                     unset
  TERM:                         xterm-ghostty
  JS_RUNTIME_VERSION:           v22.14.0
  JS_RUNTIME_NAME:              node
  NODE_PACKAGE_MANAGER:         bun/1.2.9

Biome Configuration:
  Status:                       Loaded successfully
  Path:                         /Users/mike/dev/work/notifications-monorepo/biome.jsonc
  Formatter enabled:            true
  Linter enabled:               true
  Assist enabled:               true
  VCS enabled:                  true

Workspace:
  Open Documents:               0
```

### What happened?

After upgrading from Biome 2.0.0-beta.1 to 2.0.0-beta.2, running biome lint fails with the following error:
```
Apply
Biome encountered an unexpected error

This is a bug in Biome, not an error in your code, and we would appreciate it if you could report it to https://github.com/biomejs/biome/issues/ along with the following information to help us fixing the issue.

Source Location: /Users/runner/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/node-semver-2.2.0/src/range.rs:982:18
Thread Name: biome::workspace_worker_12
Message: internal error: entered unreachable code: This should not have parsed: ~0.x.0
```

Steps to reproduce
Set up a Turborepo monorepo with Biome for linting (version 2.0.0-beta.1 works fine)
Update package.json to use Biome 2.0.0-beta.2
Run bun lint which executes bunx biome lint
Observe the error related to semver parsing

### Expected result

Expected result
Biome should successfully run the lint command without any internal errors, as it did with version 2.0.0-beta.1

### Code of Conduct

- [x] I agree to follow Biome's Code of Conduct
